adam/flask_backend/app.py

```diff
@@ -135,7 +143,116 @@ def get_scene() -> Any:
     }
 
 
+@app.route("/api/get_image", methods=["GET"])
+def get_image() -> Any:
+    if not request.args:
+        abort(HTTPStatus.BAD_REQUEST)
+
+    image_path = request.args.get("image_path", default="").strip("/")
+    image_file = DATA_DIR.joinpath(image_path).resolve()
+    if not is_relative_to(image_file, DATA_DIR):
+        abort(HTTPStatus.BAD_REQUEST)
+    print(image_file)
+    with open(image_file, "rb") as fp:
+        image_data = b64encode(fp.read()).decode("utf-8")
+    return {"image_data": image_data}
+
+
+@app.route("/api/poll_changes", methods=["GET"])
+def poll_changes() -> Any:
+    if not request.args:
+        abort(HTTPStatus.BAD_REQUEST)
+    learner = request.args.get("learner", default="")
+    training_curriculum = request.args.get("training_curriculum", default="")
+    testing_curriculum = request.args.get("testing_curriculum", default="")
+    if not learner and not training_curriculum and not testing_curriculum:
+        abort(HTTPStatus.BAD_REQUEST)
+
+    if any("/" in arg for arg in (learner, training_curriculum, testing_curriculum)):
+        abort(HTTPStatus.BAD_REQUEST)
+
+    curriculum_dir = (
+        LEARNERS_DIR
+        / learner
+        / EXPERIMENTS_DIR_NAME
+        / training_curriculum
+        / EXPERIMENTS_TESTING_DIR_NAME
+        / testing_curriculum
+    ).resolve()
+
+    if not is_relative_to(curriculum_dir, LEARNERS_DIR):
+        abort(HTTPStatus.BAD_REQUEST)
+
+    if (
+        not app.config["LEARNER_DIR_CONTENTS"].get(learner)
+        or not app.config["LEARNER_DIR_CONTENTS"][learner].get(training_curriculum)
+        or not app.config["LEARNER_DIR_CONTENTS"][learner][training_curriculum].get(
+            testing_curriculum
+        )
+        or not curriculum_dir.is_dir()
+    ):
+        return {"status": False, "message": "Invalid curriculum"}
+
+    ref_experiments = app.config["LEARNER_DIR_CONTENTS"][learner][training_curriculum][
+        testing_curriculum
+    ]
+    update_message = []
+    for situation_dir in curriculum_dir.iterdir():
+        if "situation" in situation_dir.name and situation_dir.is_dir():
+            if situation_dir.name not in ref_experiments:
+                update_message.append(
+                    f"New decode available for situation {int(situation_dir.name.rsplit('_')[-1]) + 1}"
+                )
+                ref_experiments[situation_dir.name] = set(
+                    file.name for file in situation_dir.iterdir()
+                )
+            else:
+                for file in situation_dir.iterdir():
+                    if file.name not in ref_experiments[situation_dir.name]:
+                        update_message.append(
+                            f"Updated information available for {int(situation_dir.name.rsplit('_')[-1]) + 1}: {file.name}"
+                        )
+                        ref_experiments[situation_dir.name].add(file.name)
```

Comment:
A quick note here. Currently, the polling function can return one of 2 options depending on the new information available. First, if there is an entirely new directory available, the function alerts the user that a new situation decode is available. Second, if a preexisting situation directory contains new files, the user is alerted to each individual new file available. This doesn't occur too often, but when it does, there is massive information overload (consider an action experiment test case, for instance, which has several camera views/time points where images are provided). I was considering one of four workarounds to improve visibility/clarity. 

1. Rather than provide new information, specify only the situation that is updated. 
2. Add a utility function to truncate the return info (not preferred because if multiple new decodes are available, some necessary information would be dropped)
3. With more entries, decrease the font size of the displayed results
4. Restrict the files considered based on our naming convention. The front end only cares about a handful of the files available, specifically
   - The id_rgb images (or just rgb if no id_rgb is available)
   - The stroke image
   - The stroke graph
   - The object decode

Thoughts? @spigo900 @lichtefeld

Reply:
Sounds good. The general structure for finding the associated files is present in `get_scene()`, so it's a fairly intuitive addition. On that note, however, I had a question on the way the RGB images are returned:
```py
"scene_images": [
            path.rsplit("data/", maxsplit=1)[-1]
            for path in sorted(
                chain(
                    glob(f"{experiment_dir}/rgb__[0-9]*.png"),
                    glob(f"{experiment_dir}/id_rgb_[0-9]*.png"),
                )
            )
        ],
```
This is the return value for RGB image paths in `get_scene()`. Currently, all the curricula I've worked with use the `id_rgb_[0-9]*` syntax, but I haven't seen any which contain `rgb__[0-9]*`. There are `rgb_[0-9]*` files available, but do we want to display those? Also, was there some case in the past when we used the `rgb__` naming convention?

Reply:
Yes, we do want to use `rgb_[0-9]*` files. They should be the same, name aside. I think that the `rgb__` convention was used previously because I know I've seen live demos where the ID-free images were shown.

Reply:
Ok, I've added functionality to check (4) above. Even when limiting only to those relevant files, displaying all the available updates still presented far too much information. While looking for any shorter ways to implement (1), I learned that there's a hack you can do in HTML where you can set the title of an element and it displays when you hover over the corresponding element. I limited the alert to the user to only display the situation label, and now it functions similarly to what (1) describes. There is a more formal way of implementing a hover display, but I'm not very familiar with it.

Reply:
So it shows the label and then the tooltip lists the image names or relative paths? That sounds fine. Thank you for the creative thinking.